### DISCUSSION TITLE
How to generate synthetic reverse phase b0 image using synb0-Disco?

### DISCUSSION THREAD

**Discussion by Sunita2023** [dipy/dipy#3296] [Preprocessing]
Dear DIPY members,

We are trying to run synb0-Disco to get inverse phase encoded image of DTI as given here (https://docs.dipy.org/stable/reference/dipy.nn.html#dipy.nn.synb0.Synb0). We have b0 of shape 128x128x88 and T1W image of shape 384x384x26. The synb0 requires the shape of 77x91x77 and we don't have a reference of this shape to register and resample.  

Are we doing some mistake? Which approach we should take?

We are stuck at this problem and not getting much help. 

Kindly please provide some guidance. 

Thanks for the help.
Sunita

**Comment by Sunita2023** [comment]
Guys who are familiar with TOPUP preprocessing of DTI please help. 
Another problem am facing is denoising. I tried with several methods proposed in DIPY. However all of these methods takes more than 15 mins. Is their any faster way to do denoising?

**Comment by Sunita2023** [comment]
Hi Serge,

Would you like to guide on how to generate a reverse phase DTI image?
I am surprised to see no reply for this query.  
Trying to carry out preprocessing of DTI which is not clear yet.  Your little guidance would be of grate help to me. 

Thanks a lot. 

Sincerely,
Sunita

**Reply by skoudoro** [reply]
> Another problem am facing is denoising. I tried with several methods proposed in DIPY. However all of these methods takes more than 15 mins.

On our lab, We have a very fast denoising version of patch2self.  Happy to share if we know more about you or via a collaboration. Also, in a closed PR, there is a faster version of mp-pca if you search in this repo.

> I am surprised to see no reply for this query.

No surprise, this is summer so vacation and conference season. 

> Would you like to guide on how to generate a reverse phase DTI image?

see @pjsjongsung answer below.

> Trying to carry out preprocessing of DTI which is not clear yet. Your little guidance would be of grate help to me.

can you give more details on what you are trying to do. preprocessing will depends a lot on your data.

**Comment by pjsjongsung** [comment]
Hi! Can you try the docker from the [original repo](https://github.com/MASILab/Synb0-DISCO) for the whole pipeline? It will have registering to atlas steps, which could make it work.

How many images do you have per subject and what methods have you tried in DIPY for denoising?

**Reply by Sunita2023** [reply]
Thank you so very much Jong and Serge for the responses. 
It seems that whenever I have questions, your conferences or workshops are happening :)

We could not try using original docker because we are using window. Is it possible to make it run for python? The docker code provided is for Linux. Registering with atlas is not a problem.

DTI shape: 128, 128, 88, 33
We have 33 volumes per subject.

 Patch2self: 1 min 30 sec with 'ols' model;  34 sec with 'ridge' model : Denoised image does not look so good.  
I also estimated mean and std of rms_diff which was 95 (161) for both models. 

Local PCA: 28 min 23.6 sec (for the same data): Denoised image looks really good. 

Marchenko-Pastur (mppca): 31 min 46 sec (for the same data): Denoised image is not so good visually. 

ASCM took 0.6 sec (for the same data): Denoised image looks okay. 

Please help us in carrying out synb0-Disco.

Thanks a lot and wishing you have a wonderful and memorable vacation.
  
Sincerely,
Sunita

**Reply by pjsjongsung** [reply]
Glad to hear you found a suitable denoising method! Unfortunately we do not have the pipeline for Synb0 integrated in DIPY, only the model yet. We are working on this. For the meanwhile, if you are windows 10 or later, they have something called WSL, which let's you use linux (including docker) in Windows. Something like [this](https://www.blopig.com/blog/2021/09/using-singularity-on-windows-with-wsl2/)? If you have problems, feel free to let us know and we can set up something together.

**Reply by Sunita2023** [reply]
Hi Jong,

Followed your instructions and installed WSL and ran the Synb0.  The part of the code from where the error is coming is given below. The last line says IO object could not be created. 

Please guide on how to proceed. What we need to change and run the program?
 
Thank you so much.  

Sincerely,
Sunita


```
Apply linear transform to T1
antsApplyTransforms -d 3 -i /tmp/tmp.J7rDx2RVjn/T1_norm.nii.gz -r /extra/atlases/mni_icbm152_t1_tal_nlin_asym_09c_2_5.nii.gz -n BSpline -t /tmp/tmp.J7rDx2RVjn/ANTS0GenericAffine.mat -o /tmp/tmp.J7rDx2RVjn/T1_norm_lin_atlas_2_5.nii.gz

Apply linear transform to distorted b0
antsApplyTransforms -d 3 -i /INPUTS/b0.nii.gz -r /extra/atlases/mni_icbm152_t1_tal_nlin_asym_09c_2_5.nii.gz -n BSpline -t /tmp/tmp.J7rDx2RVjn/ANTS0GenericAffine.mat -t /tmp/tmp.J7rDx2RVjn/epi_reg_d_ANTS.txt -o /tmp/tmp.J7rDx2RVjn/b0_d_lin_atlas_2_5.nii.gz
Exception caught during reference file reading

itk::ImageFileReaderException (0x3d08cf0)
Location: "unknown"
File: /home/local/VANDERBILT/blaberj/ANTS_13_FEB_2019/bin/ants/ITKv5-install/include/ITK-5.0/itkImageFileReader.hxx
Line: 139
Description:  Could not create IO object for reading file /INPUTS/b0.nii.gz
```

**Reply by pjsjongsung** [reply]
Hmm it seems that ants had some problems reading the file. Are you sure you have a file named b0.nii.gz in INPUTS directory?

**Reply by Sunita2023** [reply]
The file is named correctly and it is inside INPUTS directory.  There is one link which discusses the similar problem. Is it a problem of RAM?
https://github.com/MASILab/Synb0-DISCO/issues/68

We ran it again and got the same problem. Is there any simpler way to circumvent this problem?

Another question I have is 'acqparams.txt' file. What we should include inside this file? 
Presently following lines are included. Is it fine?
0 1 0 0.062
0 1 0 0.000

Thank you so much Jong for your guidance. 

Sincerely,
Sunita

**Reply by pjsjongsung** [reply]
Yes from what I remember acqparams.txt contains recommended parameters that should not need any change.

I don't think the problem is with RAM, because it is just reading a single b0 image. The link you mentioned is failing to create a mid-result file of the pipeline. Can you try opening the b0.nii.gz file? (dipy_horizon, ITK-snap, etc?) Does it open properly?

**Reply by Sunita2023** [reply]
There was a problem with b0.nii.gz file. It is now fixed. However, we are getting a new error which is given below.  Could you please help to fix this issue. 

Thanks again Jong for your help.  

```
`Removing job directory...
Performing inference on FOLD: 1
Performing inference on FOLD: 2
/extra/pipeline.sh: line 42:  1433 Killed                  python3.6 /extra/inference.py /OUTPUTS/T1_norm_lin_atlas_2_5.nii.gz /OUTPUTS/b0_d_lin_atlas_2_5.nii.gz /OUTPUTS/b0_u_lin_atlas_2_5_FOLD_"$i".nii.gz /extra/dual_channel_unet/num_fold_"$i"total_folds"$NUM_FOLDS"seed_1_num_epochs_100_lr_0.0001_betas\(0.9\,\ 0.999\)weight_decay_1e-05_num_epoch*.pth
/extra/pipeline.sh: line 42:  1640 Killed                  python3.6 /extra/inference.py /OUTPUTS/T1_norm_lin_atlas_2_5.nii.gz /OUTPUTS/b0_d_lin_atlas_2_5.nii.gz /OUTPUTS/b0_u_lin_atlas_2_5_FOLD_"$i".nii.gz /extra/dual_channel_unet/num_fold_"$i"total_folds"$NUM_FOLDS"seed_1_num_epochs_100_lr_0.0001_betas\(0.9\,\ 0.999\)weight_decay_1e-05_num_epoch*.pth
Performing inference on FOLD: 3
Performing inference on FOLD: 4
/extra/pipeline.sh: line 42:  1847 Killed                  python3.6 /extra/inference.py /OUTPUTS/T1_norm_lin_atlas_2_5.nii.gz /OUTPUTS/b0_d_lin_atlas_2_5.nii.gz /OUTPUTS/b0_u_lin_atlas_2_5_FOLD_"$i".nii.gz /extra/dual_channel_unet/num_fold_"$i"total_folds"$NUM_FOLDS"seed_1_num_epochs_100_lr_0.0001_betas\(0.9\,\ 0.999\)weight_decay_1e-05_num_epoch*.pth
Performing inference on FOLD: 5
/extra/pipeline.sh: line 42:  2054 Killed                  python3.6 /extra/inference.py /OUTPUTS/T1_norm_lin_atlas_2_5.nii.gz /OUTPUTS/b0_d_lin_atlas_2_5.nii.gz /OUTPUTS/b0_u_lin_atlas_2_5_FOLD_"$i".nii.gz /extra/dual_channel_unet/num_fold_"$i"total_folds"$NUM_FOLDS"seed_1_num_epochs_100_lr_0.0001_betas\(0.9\,\ 0.999\)weight_decay_1e-05_num_epoch*.pth
/extra/pipeline.sh: line 42:  2261 Killed                  python3.6 /extra/inference.py /OUTPUTS/T1_norm_lin_atlas_2_5.nii.gz /OUTPUTS/b0_d_lin_atlas_2_5.nii.gz /OUTPUTS/b0_u_lin_atlas_2_5_FOLD_"$i".nii.gz /extra/dual_channel_unet/num_fold_"$i"total_folds"$NUM_FOLDS"seed_1_num_epochs_100_lr_0.0001_betas\(0.9\,\ 0.999\)weight_decay_1e-05_num_epoch*.pth
Taking ensemble average
Image Exception : #63 :: No image files match: /OUTPUTS/b0_u_lin_atlas_2_5_FOLD_*
terminate called after throwing an instance of 'std::runtime_error'`
```

**Reply by pjsjongsung** [reply]
Glad that was resolved! For this error, it seems that the deep learning process is being killed for some reason. The owners of the Synb0 repo might have a better idea how to debug this.

**Reply by Sunita2023** [reply]
Dear Jong,

I have a question which is not directly related to this. For research purpose we are estimating fiber tracts using DIPY. During the refinement of fiber tracks, mni152 bundles are are used as a reference with certain RMSD cutoff for the refinement of the respective bundles. Are there any other metrics to determine if the generated bundles are correct?

The best way to solve the above problem is by getting directly the inverse phase encoded image as indirect way is not working. 

Thank you so much for your help.

**Reply by pjsjongsung** [reply]
I believe BUAN has a method of calculating shape similarity scores which you could use to compare between the atlas bundles and the generated bundles. @BramshQamar might be able to provide you with more information on this.

**Reply by BramshQamar** [reply]
Hi @Sunita2023 

@pjsjongsung is right! We have a shape similarity metric that ranges from 0-1. Where 0 means least similar, and 1 is most similar in shape. You can calculate shape similarity between the atlas bundle and your segmented/refined subject's bundle. Here's the tutorial: https://docs.dipy.org/stable/examples_built/streamline_analysis/bundle_shape_similarity.html#sphx-glr-examples-built-streamline-analysis-bundle-shape-similarity-py

We also have a deformable registration approach to assess the shape similarity of two bundles. You can register the atlas bundle with the subject's bundle using fully deformable BundleWarp registration. Fully deformable BundleWarp registration will try to warp the subject's bundle to look precisely like the atlas bundle. Through the deformation field, you can get the magnitude of deformations in mm. If two bundles are similar in shape, fewer deformations would be required in registration, and when two bundles are quite different, you will need more deformation. A higher value of deformation magnitude implies huge shape dissimilatory, and a lower value implies less shape dissimilatory (higher shape similarity). Here's the BundleWarp registration tutorial, and at the end of the tutorial gives an example of how to calculate this: https://docs.dipy.org/stable/examples_built/registration/bundlewarp_registration.html#sphx-glr-examples-built-registration-bundlewarp-registration-py
